### PR TITLE
Use CPU clock timeout for PVF jobs

### LINKED ISSUE #4217: Switch PVF preparation to CPU time
For controlling the execution time of preparation (and potentially execution) of PVF it seems to be better to use the CPU time and not the wall clock.

The CPU time can be obtained by either [clock_gettime](https://man7.org/linux/man-pages/man2/clock_gettime.2.html) or [getrusage](https://man7.org/linux/man-pages/man2/getrusage.2.html) for the calling process. #4123 makes the PVF compilation single-threaded, so there should be only one thread compiling or executing. There should be a watchdog thread though that wakes up from time to time (as roughly as 1 sec) and terminates the process in case it reached the deadline. That means that the controlling of timeouts shifts from the polkadot host into the worker, which is not ideal but acceptable.

Alternatively, we could leave it as is maybe just increasing the current timeout. Then, we would `wait4` on the children worker and get its `rusage` (not unlike the well known `time` program). This would make the preparation process completely single threaded and leave the control of the timeouts on the host. However, that seems more complicated in terms of implementation.

### PR DESCRIPTION
**Pull request by m-cat**
# PULL REQUEST

## Overview

The idea behind this change is that the CPU clock is less affected by load than the wall clock.

We confirmed this empirically with [an experiment](https://github.com/paritytech/polkadot/issues/4217#issuecomment-1306336250).

The hope is that this reduces unnecessary disputes.

### Changes

- The child worker now spawns a new separate thread, to monitor CPU time, whenever a job comes in.
  - The thread mostly sleeps and uses a strategy to wake up as little as possible. 
  - If the job goes over the timeout the whole child process gets killed.
- The host still has a timeout for waiting for a response from the child, but it is 4x as lenient as before.
  - This is a wall clock timeout, since there is no good way to get a child's CPU time before the child process exits.
  -  This is lenient because the wall clock can slow down up to a factor of 4 (tested on 32 threads of work on an 8-core machine).

## TODO

- [x] Add equivalent logic for PVF execution
- [x] Add tests
- [x] Small refactor of long match arm

## Issues closed

Closes #4217

### CODE REVIEW

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [inline node/core/pvf/src/prepare/worker.rs:313]: Could use a second opinion on this. My concern is a bunch of prepare jobs completing quickly and there being many waiting threads in the background. I don't see it being that big of a problem though.

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [inline node/core/pvf/src/prepare/worker.rs:367]: The data that the host expects from the child has changed. I think it's fine but wanted to double-check that this wouldn't cause breakage.

**Comment by vstakhov** [comment]
> The idea behind this change is that the CPU clock is less affected by load than the wall clock.

Well, I'd say that it is more affected by load as this clock actually measures it's own process load. Hence, if a process is sleeping (e.g. waiting on epoll) it's clock of type `CLOCK_PROCESS_CPUTIME_ID` is not running as well. On the other hand, if the task is to measure the own load, this timer might be a good thing to check (but less portable than `getrusage(RUSAGE_SELF)`).

Wall clock is really affected by cpu load but it shouldn't be that significant unless we use coarse timers (and we should not do it in this case).

**Comment by m-cat** [comment]
@vstakhov By "CPU clock is less affected by load than the wall clock" I mean the CPU clock of the process is less affected by load of the system (i.e. other processes). The load of the child process here is constant (preparing an artifact). We found that for a constant amount of work in the process, timing it with the cpu clock resulted in more consistent measurements than using the wall clock. Wall time varied up to a factor of 4 depending on load, whereas measured CPU time varied by less than a factor of 2. 

> this timer might be a good thing to check (but less portable than `getrusage(RUSAGE_SELF)`)

Do you mean as opposed to `clock_gettime`?

**Comment by s0me0ne-unkn0wn** [comment]
I love this one, it is not only about more precise measurements, in perspective it would allow us to enable more robust preparation and execution pipelines with better throughput, as CPU time does not depend on number of parallel workers as strongly as wall clock time.
My only concern is if it works equally well in different environments. Remember that we have to perform [really weird checks](https://github.com/paritytech/substrate/blob/master/client/executor/wasmtime/src/util.rs#L102) and [address some non-obvious problems](https://github.com/paritytech/substrate/blob/master/client/executor/wasmtime/src/util.rs#L151) in different environments. I think it should be evaluated in some way.

**Comment by vstakhov** [comment]
> Do you mean as opposed to `clock_gettime`?

Yes, afair, the way to make it working in osx was not very pretty:

```c
	thread_port_t thread = mach_thread_self();

	mach_msg_type_number_t count = THREAD_BASIC_INFO_COUNT;
	thread_basic_info_data_t info;
	if (thread_info (thread, THREAD_BASIC_INFO, (thread_info_t)&info, &count) != KERN_SUCCESS) {
		return -1;
	}

	res = info.user_time.seconds + info.system_time.seconds;
	res += ((double)(info.user_time.microseconds + info.system_time.microseconds)) / 1e6;
	mach_port_deallocate(mach_task_self(), thread);
```

Then I've switched to `getrusage` as it seems to be now in the basic POSIX [spec](https://pubs.opengroup.org/onlinepubs/009696699/functions/getrusage.html).

Probably the support of `CLOCK_PROCESS_CPUTIME_ID` is now fixed in the more recent osx versions though.

**Comment by vstakhov** [comment]
> By "CPU clock is less affected by load than the wall clock" I mean the CPU clock of the process is less affected by load of the system (i.e. other processes). The load of the child process here is constant (preparing an artifact). We found that for a constant amount of work in the process, timing it with the cpu clock resulted in more consistent measurements than using the wall clock.

That makes total sense, if the task is to measure own cpu load, then wall clock is not the best option due to preemption and other issues. CPU clock is also affected due to cache/tlb pollution but it is definitely a better option than a wall clock. However, if the evaluated code can wait/yeild somehow, then it might be possible to abuse this timer by performing very long sleep. On the other hand, this opportunity is covered by the fact that you also fire a wall clock timer in the parent process...

**Comment by m-cat** [comment]
> Then I've switched to `getrusage` as it seems to be now in the basic POSIX [spec](https://pubs.opengroup.org/onlinepubs/009696699/functions/getrusage.html).
> 
> Probably the support of `CLOCK_PROCESS_CPUTIME_ID` is now fixed in the more recent osx versions though.

@vstakhov Looks like `clock_gettime` is also in the POSIX spec:
https://pubs.opengroup.org/onlinepubs/009696699/functions/clock_gettime.html

According to the docs for the `cpu_time` crate, MacOS has supported `clock_gettime` since 2016:
https://github.com/tailhook/cpu-time/blob/master/src/clock_gettime.rs#L60

Not sure how old of machines we want to support. It is simpler for me to use the `cpu_time` crate but I can switch to `getrusage` if you think that would be better.

**Comment by vstakhov** [comment]
> Looks like `clock_gettime` is also in the POSIX spec:

It is in the `realtime` part that is technically an extension. But yes, it seems to be supported even by osx nowadays :)



> It is simpler for me to use the `cpu_time` crate but I can switch to `getrusage` if you think that would be better.

I think this switch might be useful merely if there are any advantages of `getrusage` that are useful in our case. In general, `getrusage` provides more information than just cpu ticks. The question is whether there is any use case of that extra information. If not, I'd also stay with `cpu_time` as it covers the vast majority of the POSIX-ish OSes presented today.

**Comment by m-cat** [comment]
> I think this switch might be useful merely if there are any advantages of `getrusage` that are useful in our case. In general, `getrusage` provides more information than just cpu ticks. The question is whether there is any use case of that extra information. If not, I'd also stay with `cpu_time` as it covers the vast majority of the POSIX-ish OSes presented today.

Makes sense. I had a look, but there doesn't seem to be anything that could be useful:

```
struct rusage {
    struct timeval ru_utime; /* user CPU time used */
    struct timeval ru_stime; /* system CPU time used */
    long   ru_maxrss;        /* maximum resident set size */
    long   ru_ixrss;         /* integral shared memory size */
    long   ru_idrss;         /* integral unshared data size */
    long   ru_isrss;         /* integral unshared stack size */
    long   ru_minflt;        /* page reclaims (soft page faults) */
    long   ru_majflt;        /* page faults (hard page faults) */
    long   ru_nswap;         /* swaps */
    long   ru_inblock;       /* block input operations */
    long   ru_oublock;       /* block output operations */
    long   ru_msgsnd;        /* IPC messages sent */
    long   ru_msgrcv;        /* IPC messages received */
    long   ru_nsignals;      /* signals received */
    long   ru_nvcsw;         /* voluntary context switches */
    long   ru_nivcsw;        /* involuntary context switches */
};
```

**Comment by m-cat** [comment]
At this point, I am just waiting for some quick feedback from @pepyakin or @slumber before finalizing this PR and moving it out of the draft stage.

File: node/core/pvf/src/prepare/worker.rs
Comment (eskimor) [reply node/core/pvf/src/prepare/worker.rs:142]: As we treat preparation as failed in this case, the answer is yes.

File: node/core/pvf/src/artifacts.rs
Comment (m-cat) [reply node/core/pvf/src/artifacts.rs:105]: I made the opposite mistake. I didn't see that the issue also mentions execution. :P

> For controlling the execution time of preparation (and potentially execution)

I think it makes sense to add the equivalent logic for execution.

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs:41]: Yeah, you're right. Another way to think about it is: if we have a 5 second timeout and the job takes 2.5 s normally, under load it can take up to 5 s CPU time and 10 s wall clock time. I will keep the conservative timeout factor.

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs:45]: I believe sleeping threads shouldn't use any CPU. It's more for thread overhead, delays in waking up (sleeps are not totally exact), etc. 

Also, this serves the double function of preventing waking up the thread for short waits. Say there is 10ms remaining, we might wait 10ms and wake it up only to find that cpu time went slower than expected and there is still 5ms. We repeat and wait 5ms, then 2ms, then 1ms. To prevent this situation we can add a bit of buffer to the sleep time.

Comment (eskimor) [reply node/core/pvf/src/prepare/worker.rs:45]: Ok, I guess that makes sense. Although why do we add it [here](https://github.com/paritytech/polkadot/blob/e34825aa290d2b05f51f62c3afc6f4c7f95f4bcc/node/core/pvf/src/prepare/worker.rs#L129)?

Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs:45]: Oh, I figured that if the job completed before the process actually terminates, it's fine, even if it took a bit longer than the timeout. But now I think we should still enforce the timeout there, where we have the exact CPU time the job took.

I still think we should always add this buffer in the child. I ran some more tests and CPU clock always seems at least slightly slower than wall clock. So if we just slept for 5 s wall clock time, only 4.99 s CPU time may have passed.

**Comment by eskimor** [comment]
> > I think this switch might be useful merely if there are any advantages of `getrusage` that are useful in our case. In general, `getrusage` provides more information than just cpu ticks. The question is whether there is any use case of that extra information. If not, I'd also stay with `cpu_time` as it covers the vast majority of the POSIX-ish OSes presented today.
> 
> Makes sense. I had a look, but there doesn't seem to be anything that could be useful:
> 
> ```
> struct rusage {
>     struct timeval ru_utime; /* user CPU time used */
>     struct timeval ru_stime; /* system CPU time used */
>     long   ru_maxrss;        /* maximum resident set size */
>     long   ru_ixrss;         /* integral shared memory size */
>     long   ru_idrss;         /* integral unshared data size */
>     long   ru_isrss;         /* integral unshared stack size */
>     long   ru_minflt;        /* page reclaims (soft page faults) */
>     long   ru_majflt;        /* page faults (hard page faults) */
>     long   ru_nswap;         /* swaps */
>     long   ru_inblock;       /* block input operations */
>     long   ru_oublock;       /* block output operations */
>     long   ru_msgsnd;        /* IPC messages sent */
>     long   ru_msgrcv;        /* IPC messages received */
>     long   ru_nsignals;      /* signals received */
>     long   ru_nvcsw;         /* voluntary context switches */
>     long   ru_nivcsw;        /* involuntary context switches */
> };
> ```

ru_maxrss ... could be interesting. By checking the amount of memory used, we might get a better heuristic on whether something is likely valid, but the machine has issues or likely invalid (excessive RAM usage). Don't have a clear strategy, seems very brittle at first thought.

Another interesting value: 

`timeval ru_stime` ... system CPU time. E.g. user cpu time is low, but system CPU time is already at 10 seconds .... something is off, kill it. 

Both are not for this PR, just values that struck me as "Could potentially be useful, maybe, some day.."

**Comment by eskimor** [comment]
> At this point, I am just waiting for some quick feedback from @pepyakin or @slumber before finalizing this PR and moving it out of the draft stage.

What feedback do you need exactly?

**Comment by eskimor** [comment]
> @eskimor Hmm, maybe worth logging at least?

Definitely useful information to have!



> Basically just whether the approach makes sense and the changes look good. I was hoping for some quick feedback before I go and write tests for this.

To me approach and changes seem sensible. Second pair of eyes would not harm though.

**Review by slumber — APPROVED** [review]
Well done! This will be very helpful if we ever decide to switch back to multithreaded preparation.

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs:310]: We have at least two threads here, don't we? If the prepare job finishes before the sleep, the sleep thread will still be sleeping in the background until it wakes up. This is because threads can't safely be killed AFAIK, so we just wait for it to kill itself on wake. But theoretically we could end up with a bunch of sleeping threads. I think it's fine because I don't think we would have too many prepare jobs at once, and also sleeping threads are not really a drain on resources.

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs:139]: It should be cleared by `with_tmp_file`, this is running in a closure provided to that function. Updated the comment.

File: node/core/pvf/src/prepare/worker.rs
Comment (slumber) [reply node/core/pvf/src/prepare/worker.rs:310]: The sleep time is small, the thread will kill itself quickly. I don't see how a threadpool would help here as the thread exists in a separate temporary process.

Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs:310]: I agree, just wanted to double-check. Will remove the comment. What do you mean by "separate temporary process"? The child prepare worker?

Comment (slumber) [reply node/core/pvf/src/prepare/worker.rs:310]: Yes

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs:335]: Good catch on the atomic bool. I made the change, but I think the `unreachable!` call can still stay, as I decided to use the strictest memory access on the atomic bool. I also considered `never!` but I think it only panics in debug mode, whereas here we have a proof that it really is unreachable. I'm not totally sure when to use `never!` vs. `unreachable!` though.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2065634

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs:209]: That's true. I didn't think it was necessary to have two different `TimedOut` err types. Can always address as part of https://github.com/paritytech/polkadot/issues/3754 if someone feels otherwise.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2065757

File: node/core/pvf/src/prepare/worker.rs
Comment (slumber) [inline node/core/pvf/src/prepare/worker.rs:177]: nit: this line belongs to the `Ok(Ok(...))` match arm to explain what was meant by nested result.

File: node/core/pvf/src/prepare/worker.rs
Comment (slumber) [inline node/core/pvf/src/prepare/worker.rs:169]: This is a little deceiving. It doesn't handle receiving, it decoded an already received bytes.

File: node/core/pvf/src/prepare/worker.rs
Comment (slumber) [inline node/core/pvf/src/prepare/worker.rs:370]: This held true when you had a mutex, now it's not. Suppose a sleeping thread wakes up and detects a timeout, it stores `true` in the atomic flag and OS scheduler puts it to a short sleep giving a window for execution of this part of code.

Previously you were holding a locked mutex here, now you don't and there's a panic (although harmless).
Instead, you should join here on a thread handle, giving it time to send error through the socket, then return.

File: node/core/pvf/src/prepare/worker.rs
Comment (slumber) [inline node/core/pvf/src/prepare/worker.rs:361]: nit: memory orderings only matter when you have a memory access to shared resources. Since there're no accesses to reorder, all can be changed to `Relaxed`.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2066275

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs:354]: Oh, is the `AtomicBool` not a shared value in this case? I mostly based the decision on this snippet:

> Relaxed operations are appropriate for things that you definitely want to happen, but don't particularly otherwise care about. For instance, incrementing a counter can be safely done by multiple threads using a relaxed fetch_add if you're not using the counter to synchronize any other accesses.

From [here](http://doc.rust-lang.org/1.65.0/nomicon/atomics.html). Since we are using the value to synchronize accesses I wanted to go with a stronger ordering. Please let me know if I'm misunderstanding.

Comment (slumber) [reply node/core/pvf/src/prepare/worker.rs:354]: You're misunderstanding the purpose of memory orderings. The atomic flag itself is safe to be shared, compare-and-exchange is atomic CPU instruction and once one thread reads-modifies-writes, subsequent read will observe it.

> Memory orderings specify the way atomic operations synchronize memory. In its weakest [Ordering::Relaxed](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.Relaxed), only the memory _**directly touched**_ by the operation is synchronized. On the other hand, a store-load pair of [Ordering::SeqCst](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.SeqCst) operations synchronize _**other**_ memory while additionally preserving a total order of such operations across all threads.

For reference, you can take a look at implementation of futex in Rust std, how it puts `Acquire` on the lock and `Release` on unlock (and why).

Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs]: Ah that makes sense, thanks for explaining!

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs]: You're right, I missed that. I tried to rework this part, let me know how it looks.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2066557

File: node/core/pvf/src/prepare/worker.rs
Comment (slumber) [inline node/core/pvf/src/prepare/worker.rs:379]: Why did you decide to move this code into `prepare_artifact`? To me it clearly belongs to some worker-related code, now it's unclear what `prepare_artifact` is responsible for (ditto for execute part).

File: node/core/pvf/src/prepare/worker.rs
Comment (slumber) [inline node/core/pvf/src/prepare/worker.rs:387]: This unreachable looks really scary and unnecessary, I would've went with a simple
```rust
// Monitor thread detected timeout and likely already terminated a process, nothing to do.
return
```

**Review by slumber — APPROVED** [review]
2 more nits from me, other than that good to go

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs:379]: Yeah, I wasn't sure about this change. In the execute part we were already getting the elapsed duration, which is a necessary part of the response returned from that function. And I wanted to put the lock flag as close as possible to the end of the execute operation, so I stuck it in the middle of that function kind of awkwardly. Then I figured I could do the same in the prepare part, for parity with the execute code and also to reduce indentation (I believe it saved 4 indent levels).

I considered alternatively adding another function wrapper around `prepare_artifact` called `worker_prepare_job` or something, which would have all this thread coordination logic in it. But I feel like it'd also be awkward to have this big chain of function calls.

I thought the present solution wasn't too bad, because `prepare_artifact` was not doing much previously -- it was basically just two function calls, to `prevalidate` and `prepare`. Felt like the least annoying option.

Comment (slumber) [reply node/core/pvf/src/prepare/worker.rs:379]: I would've still went with [this](https://github.com/paritytech/polkadot/blob/8e5139f1224eb1bd1c0caa1221fca6230beff583/node/core/pvf/src/prepare/worker.rs) implementation, (personally) the readability was better despite indentation levels

File: node/core/pvf/src/prepare/worker.rs
Comment (eskimor) [reply node/core/pvf/src/prepare/worker.rs:387]: We can also have a log or even a debug_assert.

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs:379]: Okay, I'll change it back. Thanks for the feedback.

File: node/core/pvf/src/prepare/worker.rs
Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs:387]: Hmm, I don't see the point of logging a case that's unreachable, which we have a proof for. I would personally keep the unreachable and maybe just make it less verbose. We also can't just do an empty return here since `worker_event_loop` expects an `io::Result<Never>>`.

Comment (slumber) [reply node/core/pvf/src/prepare/worker.rs:387]: Yes, `continue` instead. Or
> make it less verbose

would also work

Comment (m-cat) [reply node/core/pvf/src/prepare/worker.rs]: Makes sense, removed the `unreachable!`.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2068563

**Comment by m-cat** [comment]
Looks like the existing tests already cover timeouts, as well as successful execution within the timeout. I added some timing asserts where I could, but I wasn't sure how to test the CPU-time-specific changes, or if we need to.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2074091

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2079628

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2087671

File: node/core/pvf/src/execute/worker.rs
Comment (eskimor) [inline node/core/pvf/src/execute/worker.rs:154]: I actually don't understand this comment. This looks like a "hard" timeout to me.

**Review by eskimor — APPROVED** [review]
Nice work @m-cat !

File: node/core/pvf/src/execute/worker.rs
Comment (m-cat) [reply node/core/pvf/src/execute/worker.rs:154]: It was maybe poorly explained or wrong choice of words, but my thinking here was that the hard timeout is on the child, whereas here on the host it is a lenient timeout. (The comment is specifically referring to the `HardTimeout` error we return further down.) The distinction matters to me because reaching the lenient timeout may indicate that the child stalled, which we may want to treat differently. At any rate, this is something I will revisit soon when I get to the linked issue.

**Pull request merged**
